tensorflow/core/kernels/dml_crop_and_resize_grad_image_op.cc

```diff
@@ -0,0 +1,217 @@
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/bounds_check.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class CropAndResizeGradImageInitHelper : public InitializationHelper {
+ public:
+  struct Attributes {
+    explicit Attributes(OpKernelConstruction* ctx) {
+      std::string method_attr;
+
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("method", &method_attr));
+      OP_REQUIRES(ctx, method_attr == "bilinear" || method_attr == "nearest",
+                  errors::InvalidArgument(
+                      "method must be 'bilinear' or 'nearest'", method_attr));
+
+      interpolation_mode = method_attr == "bilinear"
+                               ? DML_INTERPOLATION_MODE_LINEAR
+                               : DML_INTERPOLATION_MODE_NEAREST_NEIGHBOR;
+    }
+
+    DML_INTERPOLATION_MODE interpolation_mode;
+  };
+
+  CropAndResizeGradImageInitHelper(OpKernelContext* ctx,
+                                   std::shared_ptr<const Attributes> attr)
+      : attr_(std::move(attr)) {
+    // The shape of 'grads' is [num_boxes, crop_height, crop_width, depth].
+    const Tensor& grads = ctx->input(0);
+    // The shape of 'boxes' is [num_boxes, 4].
+    const Tensor& boxes = ctx->input(1);
+    // The shape of 'box_index' is [num_boxes].
+    const Tensor& box_index = ctx->input(2);
+    // The shape of 'image_size' is [4].
+    const Tensor& image_size = ctx->input(3);
+
+    // Validate input shapes.
+    OP_REQUIRES(ctx, grads.dims() == 4,
+                errors::InvalidArgument("grads image must be 4-D",
+                                        grads.shape().DebugString()));
+    const int crop_height = grads.dim_size(1);
+    const int crop_width = grads.dim_size(2);
+
+    OP_REQUIRES(ctx, crop_height > 0 && crop_width > 0,
+                errors::InvalidArgument("grads dimensions must be positive"));
+
+    int num_boxes = 0;
+    const TensorShape& boxes_shape = ctx->input(1).shape();
+    const TensorShape& box_index_shape = ctx->input(2).shape();
+
+    if (boxes_shape.num_elements() != 0 || boxes_shape.num_elements() != 0) {
```

Comment:
What's the difference between `boxes_shape.num_elements()` and `boxes_shape.num_elements()`?